core/src/client.rs

```diff
@@ -0,0 +1,231 @@
+			Err(error) => Err(RpcError::Other(error.into())),
+		}
+	}
+}
+
+/// A message sent to the `RpcClient`. This is public so that
+/// the derive crate can generate a client.
+pub struct RpcMessage {
+	/// The rpc method name.
+	pub method: String,
+	/// The rpc method parameters.
+	pub params: Params,
+	/// The oneshot channel to send the result of the rpc
+	/// call to.
+	pub sender: oneshot::Sender<Result<Value, Error>>,
+}
+
+/// A channel to a `RpcClient`.
+pub type RpcChannel = mpsc::Sender<RpcMessage>;
+
+/// The RpcClient handles sending and receiving asynchronous
+/// messages through an underlying transport.
+pub struct RpcClient<TSink, TStream> {
+	id: u64,
+	queue: HashMap<Id, oneshot::Sender<Result<Value, Error>>>,
+	sink: TSink,
+	stream: TStream,
+	ch: mpsc::Receiver<RpcMessage>
+}
+
+impl<TSink, TStream> RpcClient<TSink, TStream> {
+	/// Creates a new `RpcClient`.
+	pub fn new(
+		sink: TSink,
+		stream: TStream,
+		ch: mpsc::Receiver<RpcMessage>,
+	) -> Self {
+		RpcClient {
+			id: 0,
+			queue: HashMap::new(),
+			sink,
+			stream,
+			ch,
+		}
+	}
+
+	fn next_id(&mut self) -> Id {
+		let id = self.id;
+		self.id = id + 1;
+		Id::Num(id)
+	}
+}
+
+impl<TSink, TStream> Future for RpcClient<TSink, TStream>
+where
+	TSink: Sink<SinkItem=String, SinkError=RpcError>,
+	TStream: Stream<Item=String, Error=RpcError>,
+{
+	type Item = ();
+	type Error = RpcError;
+
+	fn poll(&mut self) -> Result<Async<Self::Item>, Self::Error> {
+		// Handle requests from the client.
+		loop {
+			let msg = match self.ch.poll() {
+				Ok(Async::Ready(Some(msg))) => msg,
+				Ok(Async::Ready(None)) => {
+					// When the channel is dropped we still need to finish
+					// outstanding requests.
+					break;
+				},
+				Ok(Async::NotReady) => break,
+				Err(()) => continue,
+			};
+			let id = self.next_id();
+			let request = Request::Single(Call::MethodCall(MethodCall {
+				jsonrpc: Some(Version::V2),
+				method: msg.method,
+				params: msg.params,
+				id: id.clone(),
+			}));
+			self.queue.insert(id, msg.sender);
+			let request_str = serde_json::to_string(&request)
+				.map_err(|error| RpcError::Other(error.into()))?;
+			self.sink.start_send(request_str)?;
+		}
+		// Handle outgoing rpc requests.
+		self.sink.poll_complete()?;
+		// Handle incoming rpc requests.
+		loop {
+			let response_str = match self.stream.poll() {
+				Ok(Async::Ready(Some(response_str))) => response_str,
+				Ok(Async::Ready(None)) => {
+					// The websocket connection was closed so the client
+					// can be shutdown. Reopening closed connections must
+					// be handled by the transport.
+					return Ok(Async::Ready(()))
+				},
+				Ok(Async::NotReady) => break,
+				Err(err) => Err(err)?,
+			};
+			let response = serde_json::from_str::<Response>(&response_str)
+				.map_err(|error| RpcError::Other(error.into()))?;
+			let outputs: Vec<Output> = match response {
+				Response::Single(output) => vec![output],
+				Response::Batch(outputs) => outputs,
```

Comment:
We never send a `Batch` request so this branch should just trigger a warning an be ignored.

Reply:
Can't a different jsonrpc server implementation return a batch response?

Reply:
The server should only respond with `Batch` if client sends a batch request: https://www.jsonrpc.org/specification#batch

If you prefer to leave as is I'm also fine, but I think not handling `Batch` would be a bit more strict and would simplify the code tiny bit.